src/shims/foreign_items.rs

```diff
@@ -111,25 +115,190 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
+                // <method_ptr_n> (usize)
+                //```
+                //
+                // Since box_me_up is the first method pointer
+                // in the vtable, we use an offset of 3 pointer sizes
+                // (skipping over <drop_ptr>, <size>, and <align>)
+
+                /*let box_me_up_ptr = alloc.read_ptr_sized(
+                    this,
+                    vtable_ptr.offset(this.pointer_size() * 3, this)?
+                )?.to_ptr()?;*/
+
+                // Get the actual function instance
+                //let box_me_up_fn = this.memory().get_fn(box_me_up_ptr)?;
+                let box_me_up_mir = this.load_mir(box_me_up_fn.def, None)?;
+
+                // Extract the signature
+                // We know that there are no HRBTs here, so it's fine to use
+                // skip_binder
+                let fn_sig_temp = box_me_up_fn.ty(*this.tcx).fn_sig(*this.tcx);
+                let fn_sig = fn_sig_temp.skip_binder();
+
+                // This is the layout of '*mut (dyn Any + Send)', which
+                // is the return type of 'box_me_up'
+                let dyn_ptr_layout = this.layout_of(fn_sig.output())?;
+
+                // We allocate space to store the return value of box_me_up:
+                // '*mut (dyn Any + Send)', which is a fat 
+                
+                let temp_ptr = this.allocate(dyn_ptr_layout, MiriMemoryKind::UnwindHelper.into());
+                this.machine.box_me_up_tmp_ptr = Some(temp_ptr.clone());
+
+                this.push_stack_frame(
+                    box_me_up_fn,
+                    box_me_up_mir.span,
+                    box_me_up_mir,
+                    Some(temp_ptr.into()),
+                    // We know that this frame will always be unwound,
+                    // since we enable unwinding as soon as we encounter it
+                    // in Evaluator::stack_pop
+                    StackPopCleanup::Goto { ret, unwind }
+                )?;
+
+                // This flag is checked by 'MiriEvalContext::stack_pop',
+                // and is used to determine when we've finished execution of
+                // 'box_me_up'
+                this.frame_mut().extra.is_box_me_frame = true;
+
+                let mut args = this.frame().body.args_iter();
+                let arg_0 = this.eval_place(&mir::Place {
+                    base: mir::PlaceBase::Local(args.next().unwrap()),
+                    projection: Box::new([])
+                })?;
+                this.write_scalar(data_ptr, arg_0)?;
+
+                return Ok(())
+
+            }
+            // Normally, this gets resolved to the '#[panic_handler]` function
+            // during compilation. We manually forward to the panic_impl lang item,
+            // which corresponds to the function with the `#[panic_handler]` attribute
+            //
+            // This is used by libcore to forward panics to the actual
+            // panic impl
```

Comment:
If we are forwarding, why can't we return that functions MIR? The shims handling is all called from [`find_fn`](https://github.com/rust-lang/miri/blob/cfd95998d7e4a24672d0b16e71e6e8dd1fdf3dd7/src/machine.rs#L171), so if we can find that MIR somewhere we should do that and not push stack frames ourselves.

Reply:
Oh wait, you have some logic for that in `shims/mod.rs`, but then why is this here still needed?

Reply:
This code is responsible for handling the really weird linkage of `panic_impl`. During a normal build, this gets resolved at link time, using some strange linker stuff.

I chose to handle it here, since it relies on matching the link name like the other things in this file.

Reply:
What I am wondering is, how does this relate to https://github.com/rust-lang/miri/pull/693/files#diff-7468b23c19621ccc2ad61dc710e27f8dR34 ? Both end up calling `this.tcx.lang_items().panic_impl()` but in totally different ways. That's odd.

Reply:
This was due to the fact that `emulate_foreign_item` is expected to completely handle the foriegn item - it doesn't return a `Body`. However, it would probably be simpler to return an `Option<Body>` to allow the normal function call logic to be re-used here.

Reply:
No that was not my question. My question is, isn't this dead code because the other code I linked to already handles calling the lang item?

Reply:
Hmm - I thought that both were necessary, but it turns out that only this one is. I've removed the other case.